### ISSUE TITLE
Build a causal observation-action-effect trace for debugging, learning and evaluation

### ISSUE THREAD

**Issue by Unjuno** [Unjuno/agent-interface#15]
## Motivation

Agent Interface increasingly needs to answer questions such as:

- Which observation caused this action?
- Which authority grant allowed it?
- Which local correction produced the visible effect?
- Which planner boundary consumed most of the time?
- Which training example came from Astra, a deterministic controller, an oracle or a learned policy?
- Did a failure originate in perception, admission, execution, application response or evaluation?

Today this information exists across event logs, screenshots, receipts, owner records, manifests and benchmark reports, but not as one explicit causal lineage.

As the project adds local servo, reactive macros, background work, speculative futures and learned accelerators, a **causal trace** becomes increasingly valuable for correctness analysis and for generating trustworthy training/evaluation data.

This issue proposes a compact implementation-independent trace contract. It is not a request to persist every raw screenshot forever.

Related issues:

- #3/#7/#8 — presentation provenance
- #5 — observation/effect feedback
- #9 — Astra-generated training data and planner-latency telemetry
- #10 — endpoint timing / Python versus Rust comparison
- #11 — shared ABI / failure locus / version identity
- #12 — benchmark/oracle evidence
- #13 — speculative futures
- #14 — retained interrupts/events

## Core idea

Represent execution as a causally linked graph rather than only a chronological log.

Conceptually:

```text
Observation
    ↓ supports
PlannerDecision / LocalDecision
    ↓ proposes
Intent / Program / Future
    ↓ admitted_by
Authority
    ↓ executes
InputAction
    ↓ followed_by
Observation / Event
    ↓ measures
Effect
    ↓ verified_by
SemanticVerifier / TaskOracle
```

Chronological timestamps remain important, but causality should be explicit rather than inferred only from temporal proximity.

## Candidate identifiers

Reuse or introduce stable IDs such as:

```text
trace_id
observation_id
presentation_id
planner_decision_id
intent_id
program_id
future_id
authority_id
owner_id / owner_revision
action_id
event_id
effect_id
evaluation_id
model/policy/version identity
```

Not every component needs every ID. The goal is sufficient linkage to reconstruct the declared causal path.

## 1. Observation provenance

For every model-visible or local-controller-visible observation, retain lineage such as:

```text
source capture
→ reconstruction/transport
→ crop/grid/composite transform
→ delivered artifact
```

A presentation transform should point back to the source observation and coordinate mapping.

This helps distinguish:

- capture failure;
- transport failure;
- presentation error;
- model perception error;
- stale observation use.

## 2. Decision provenance

A decision record should say what produced it without requiring hidden reasoning traces.

Examples:

```text
producer = Astra
producer = deterministic_macro
producer = patch_servo
producer = learned_policy:<version>
producer = human/test_fixture
```

Record the observable inputs/references the decision was allowed to use.

Do not depend on private chain-of-thought.

## 3. Authority lineage

Every executed input should be attributable to the authority that admitted it.

Useful links include:

```text
action
→ authority_id
→ intent_id
→ validity dependencies
→ lease deadline
```

Cleanup/release actions should be distinguishable from new task input after expiry.

This makes it possible to audit that no action executed using a released, obsolete or unrelated authority.

## 4. Effect lineage

Do not stop at input execution.

Link subsequent local measurements to the actions they are intended to evaluate:

```text
pointer move
→ anchor displacement
→ local goal result
→ semantic verifier
→ task oracle
```

Multiple actions may contribute to one effect, and one action may have multiple observed effects.

The trace should therefore allow many-to-many references rather than assuming one action = one effect.

## 5. Event causality

Events should record relevant causes/predecessors where known.

Examples:

```text
TARGET_LOST caused local servo stop
LEASE_EXPIRED caused independent release
FOCUS_LOST invalidated authority
GOAL_REACHED superseded progress state
```

Do not invent causality when only temporal ordering is known. Use explicit distinctions such as:

```text
caused_by
observed_after
related_to
```

where appropriate.

## 6. Planner-boundary telemetry

A causal trace should make the current timing questions easier to answer.

Record explicit endpoints such as:

```text
observation ready
observation delivered
planner reply received
program received
program accepted
first input
first effect
terminal
semantic verification
```

Then calculate boundary gaps from IDs/endpoints rather than ad hoc log searches.

This supports #9 planner-availability profiles and #10 implementation-overhead separation.

## 7. Training-data provenance

For ephemeral learned control, every training item should identify its supervision source.

Example:

```text
TrainingExample {
    source_observation_ids
    goal/intent_id
    action labels
    effect/evaluation refs
    label_source
    augmentation_parent
    policy_version_if_generated
}
```

Candidate label sources:

```text
ASTRA_DEMONSTRATION
ASTRA_CORRECTION
INDEPENDENT_ORACLE
DETERMINISTIC_CONTROLLER
SYNTHETIC_TRANSFORM
LEARNED_POLICY_UNVERIFIED
```

The last category should not silently become ground truth.

This makes self-training contamination and train/eval leakage auditable.

## 8. Benchmark provenance

A benchmark result should link to:

- task contract/scorer version;
- initial-state identity;
- controller/model/runtime versions;
- observation/action/effect trace;
- final independent evaluation;
- failure locus/classification if unsuccessful.

This helps preserve negative cohorts and compare candidates without losing which semantics actually ran.

## 9. Speculative future provenance

For #13-style futures, record:

```text
prepared_from
expected_predicate
matched_by_observation/event
admission_result
executed_program OR discard_reason
```

A discarded future remains useful evidence when measuring speculation hit rate and wasted planner work.

## 10. Compactness and retention

Do not require indefinite storage of all raw pixels.

Separate:

```text
causal metadata
artifact hashes/references
raw artifacts
```

Possible retention modes:

### Ephemeral

Raw screenshots/training examples deleted after session; causal metadata and aggregate metrics may remain.

### Frozen research cohort

Selected artifacts retained with hashes for reproducibility/audit.

### Minimal production mode

Only bounded operational metadata needed for correctness/recovery.

Privacy and data-retention policy should be explicit. A trace system is not justification to collect more user data than needed.

## 11. Replay and differential analysis

The causal trace should support useful offline analyses without pretending historical replay is a live environment.

Examples:

- rerun a visual detector on frozen source images;
- compare Python and Rust admission decisions on the same recorded state;
- run a learned policy in shadow mode on historical observations;
- recompute presentation variants;
- verify event/completion transition logic;
- compare candidate decisions against the recorded independent task outcome.

Do not claim interactive success from replay when the controller could not affect future frames.

## 12. Failure diagnosis

A trace should make it possible to distinguish paths such as:

```text
correct observation
→ wrong planner action
```

versus

```text
correct action
→ stale authority rejection
```

versus

```text
input executed
→ application effect absent
```

versus

```text
application effect correct
→ evaluator incorrect
```

This connects naturally to `failure_locus` in #11.

## 13. Integrity

For frozen experiments, use hashes/manifests where practical so later analysis knows which exact artifacts and code versions produced the evidence.

The trace itself should not claim cryptographic security merely because hashes exist. The goal is reproducible research provenance, not remote attestation unless separately designed.

## Evaluation

The trace layer is useful only if its overhead and debugging value justify it.

Measure:

- additional local serialization/storage cost;
- event volume;
- ability to derive planner-boundary timings automatically;
- ability to reproduce current audits with less bespoke code;
- ability to detect deliberately injected stale/misattributed actions;
- train/eval provenance coverage;
- missing-link/error rate in frozen cohorts.

Do not put high-rate verbose tracing on the critical input path by default if it measurably harms cadence. Buffering or sampled detail may be appropriate outside frozen research mode.

## First experiment

Take one existing actual-assistant servo/recovery cohort and encode its existing evidence into the proposed causal graph without changing behavior.

Ask whether the graph can answer, mechanically:

1. which model-visible image supported each planner action;
2. which authority admitted each input;
3. which local feedback caused each continuation/stop;
4. where the dominant planner gaps occurred;
5. which independent evidence proves final task success;
6. which failure evidence would be retained if the task failed.

If the model requires ad hoc log interpretation for these questions, refine the trace contract before adding more instrumentation.

## Design principle

> Every important action should have a visible reason, authority and effect lineage.

As Agent Interface becomes more autonomous and more adaptive, causal provenance is what lets us distinguish useful acceleration from hidden behavior, build clean training data, compare implementations and preserve failed evidence without guessing after the fact.
